4932 - Form improvements: Phone input field with prefix picker

sorja opened the pull request:
https://github.com/user-attachments/assets/39818ed2-b0ed-4c79-99c4-fff1e49ae66d

sorja committed:
4932 - Areas: move to ./areas, introduce getCountryIso2 and getCountryIsoFromCallingCode

sorja commented:
@minotogna If you're curious:

Here is sneak peek on the telephone field I am working on
I tried different ways, but I think this looks pretty neat and sits well in the form

It's only using 2 small packages, some helper to convert ISO3 -> ISO2 and to get the calling code (only supports ISO2)
Later on if we decide to add ISO2 to country prop, we can remove the other package :)

Note: The telephone number is only used for viewing, so it shouldn't matter in what format we save it in database (free text)

minotogna commented:
> @minotogna If you're curious:
> 
> 
> 
> Here is sneak peek on the telephone field I am working on
> 
> I tried different ways, but I think this looks pretty neat and sits well in the form
> 
> 
> 
> It's only using 2 small packages, some helper to convert ISO3 -> ISO2 and to get the calling code (only supports ISO2)
> 
> Later on if we decide to add ISO2 to country prop, we can remove the other package :)
> 
> 
> 
> Note: The telephone number is only used for viewing, so it shouldn't matter in what format we save it in database (free text)

That's nice @sorja 👍
Regarding iso 2, do u suggest to save it in the db ?

sorja commented:
> > @minotogna If you're curious:
> > Here is sneak peek on the telephone field I am working on
> > I tried different ways, but I think this looks pretty neat and sits well in the form
> > It's only using 2 small packages, some helper to convert ISO3 -> ISO2 and to get the calling code (only supports ISO2)
> > Later on if we decide to add ISO2 to country prop, we can remove the other package :)
> > Note: The telephone number is only used for viewing, so it shouldn't matter in what format we save it in database (free text)
> 
> That's nice @sorja 👍 Regarding iso 2, do u suggest to save it in the db ?

We can do that too, but maybe not now if we only use it here?
I  might be mistaken - but it seems that ISO2 is more commonly used?

minotogna commented:
> > > @minotogna If you're curious:
> 
> > > Here is sneak peek on the telephone field I am working on
> 
> > > I tried different ways, but I think this looks pretty neat and sits well in the form
> 
> > > It's only using 2 small packages, some helper to convert ISO3 -> ISO2 and to get the calling code (only supports ISO2)
> 
> > > Later on if we decide to add ISO2 to country prop, we can remove the other package :)
> 
> > > Note: The telephone number is only used for viewing, so it shouldn't matter in what format we save it in database (free text)
> 
> > 
> 
> > That's nice @sorja 👍 Regarding iso 2, do u suggest to save it in the db ?
> 
> 
> 
> We can do that too, but maybe not now if we only use it here?
> 
> I  might be mistaken - but it seems that ISO2 is more commonly used?

@sorja i honestly don't know what is more used . However, if it doesn't cost too much, we can add the iso 2 code in the db now . That would be a better solution . What do you think ?

sorja committed:
4932 - migration-step template: return type

sorja committed:
4932 - add country_iso_2

sorja committed:
4932 - migration: add calling code to country

sorja commented:
> > > > @minotogna If you're curious:
> > 
> > 
> > > > Here is sneak peek on the telephone field I am working on
> > 
> > 
> > > > I tried different ways, but I think this looks pretty neat and sits well in the form
> > 
> > 
> > > > It's only using 2 small packages, some helper to convert ISO3 -> ISO2 and to get the calling code (only supports ISO2)
> > 
> > 
> > > > Later on if we decide to add ISO2 to country prop, we can remove the other package :)
> > 
> > 
> > > > Note: The telephone number is only used for viewing, so it shouldn't matter in what format we save it in database (free text)
> > 
> > 
> > > 
> > 
> > 
> > > That's nice @sorja 👍 Regarding iso 2, do u suggest to save it in the db ?
> > 
> > 
> > We can do that too, but maybe not now if we only use it here?
> > I  might be mistaken - but it seems that ISO2 is more commonly used?
> 
> @sorja i honestly don't know what is more used . However, if it doesn't cost too much, we can add the iso 2 code in the db now . That would be a better solution . What do you think ?

I started working on this and migrated the iso2 to DB. Everything worked well. I added a csv in migrations/data of format: [iso3][iso2]

After migration I realized, that we don't need the iso2 but we need the _calling code_.

So I left the csv with iso2, but added calling code: [iso3][iso2][callingcode] - so if we want the iso2 in the future, it's easy to migrate.

Now that I have calling code, I think I can get rid of both NPM packages.

Confusing message, but in short: instead of iso2, I added calling code. I left iso2 for future in case

Let me know

sorja committed:
4932 - migration: add calling code to country

sorja committed:
4932 - revert meta/areas

sorja committed:
4932 - Component: FlagIcon

sorja committed:
4932 - Component: CallingCode

sorja committed:
4932 - remove unused packages

minotogna commented:
@sorja in short, let's add iso 2 and calling code in db , ok ?

sorja committed:
4932 - migrate iso2

sorja committed:
4932 - separate field for calling code

sorja committed:
4932 - update useHandleCountryChange

sorja committed:
4932 - Component: PhoneNumberInput

sorja committed:
4932 - Component: TelephoneField

sorja committed:
4932 - User forms: use TelephoneField

sorja committed:
4932 - Less strict telephone validation

sorja committed:
4932 - Allow submitting form when phone number changes

sorja commented:
@minotogna Video updated - let me know

Developers note:
<img width="254" height="317" alt="image" src="https://github.com/user-attachments/assets/12162590-98a5-4dc3-bfb9-f1734484d535" />

This was slightly more complicated to get to work nicely
Tried different combinations, but I think the current video looks the best - just need to refine a bit how we show the two fields next to each other (e.g. separator `-`, spacing, width etc)

I also simplified the telephone number vlaidation requirements: I noticed some internal users (e.g. @ FAO ) use internal phone numbers (e.g. 061234) - I don't think there is a real reason to strictly validate the numbers, they are only for reading anyway.

I divided the component in meaningful parts:

The main component `TelephoneField` is divided in two components:
 - CallingCode (more complex logic)
 - PhoneNumberInput (simple text input)

The parent handles state for calling code and phone number


The complex component `CallingCode`:
- Contains a selector and text input
- Has some logic to glue everything together
- Contains the subcomponents:
  - Option (used to display the flag, country name and calling code in drop down)
  - FlagIcon (used by Option and CallingCode)

sorja committed:
4932 - Display irregular phone numbers correctly

sorja committed:
4932 - Align items vertically

sorja committed:
4932 - Vertical divider after calling code

sorja committed:
Merge branch 'development' into 4932-form-improvements-phone-input-field-with-prefix-picker

sorja commented:
@minotogna PR Description updated with latest video

Note after last video/changes: 
Added vertical divider after calling code
Aligned text vertically to center

minotogna commented on src/client/components/Inputs/Select/types.ts:
```diff
@@ -6,19 +7,25 @@ export type Option = {
   value: string
 }
 
+export type OptionElement = {
```
This seems a duplicate of Option . Can u use React.Node for label Option ?

minotogna commented on src/client/pages/User/hooks/useRoleValidationSchema.ts:
```diff
@@ -42,12 +42,13 @@ export const useRoleValidationSchema = () => {
-              error: t('form.errors.invalid', { field: t('editUser.primaryPhoneNumber') }),
-            }),
+            .min(1, { error: t('form.errors.required', { field: t('editUser.primaryPhoneNumber') }) }),
+          // .refine((val) => z.e164().safeParse(val).success, {
```
Should we remove the comments ?

sorja replied:
Sure, left in case we want to enable in future, but I can remove

minotogna commented on src/tools/migrations/steps/steps/data/iso3-iso2-calling_code.csv:
```diff
@@ -0,0 +1,257 @@
+iso3,iso2,calling_code
```
Should this be part of the migration step directly so that after reset it gets removed ?

sorja committed:
4932 - Remove comments

sorja committed:
4932 - Remove reduntant type OptionElement

sorja committed:
4932 - Merge data to migration step

sorja committed:
4932 - Remove atlantis from calling code selector

sorja commented:
In addition, removed Atlantis countries from calling code selector

sorja committed:
Merge branch '4932-form-improvements-phone-input-field-with-prefix-picker' of github.com:openforis/fra-platform into 4932-form-improvements-phone-input-field-with-prefix-picker

minotogna commented on src/client/components/Inputs/Select/types.ts:
```diff
@@ -2,7 +2,7 @@ import { ActionMeta, Props as ReactSelectProps } from 'react-select'
 
 export type Option = {
-  label: string
+  label: string | React.ReactNode
```
string can be removed

sorja committed:
4932 - Remove redundant type string

sorja committed:
Merge branch 'development' into 4932-form-improvements-phone-input-field-with-prefix-picker
